src/main/java/fi/aalto/cs/apluscourses/ui/exercise/ExercisesView.java

```diff
@@ -1,26 +1,53 @@
 package fi.aalto.cs.apluscourses.ui.exercise;
 
+import static fi.aalto.cs.apluscourses.utils.PluginResourceBundle.getText;
+
 import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.application.ModalityState;
 import fi.aalto.cs.apluscourses.intellij.actions.ActionUtil;
 import fi.aalto.cs.apluscourses.intellij.actions.OpenSubmissionAction;
+import fi.aalto.cs.apluscourses.presentation.MainViewModel;
 import fi.aalto.cs.apluscourses.presentation.exercise.ExercisesTreeViewModel;
 import fi.aalto.cs.apluscourses.ui.GuiObject;
 import fi.aalto.cs.apluscourses.ui.base.TreeView;
+
+import java.awt.CardLayout;
+
+import javax.swing.JLabel;
 import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.SwingConstants;
+
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+
 public class ExercisesView {
   private TreeView exerciseGroupsTree;
+  private JLabel emptyText;
   private JPanel basePanel;
-
+  private JScrollPane pane;
   @GuiObject
   public JPanel toolbarContainer;
+  private JPanel cardPanel;
+  private CardLayout cl;
+  private MainViewModel mainViewModel;
 
-  public ExercisesView() {
-    // See ModulesView.java
+  /**
+   * Creates an ExerciseView that uses mainViewModel to dynamically adjust its UI components.
+   */
+  public ExercisesView(MainViewModel mainViewModel) {
+    this.mainViewModel = mainViewModel;
     basePanel.putClientProperty(ExercisesView.class.getName(), this);
+    cl = (CardLayout) cardPanel.getLayout();
+    exerciseGroupsTree.getEmptyText().appendLine(getText("ui.exercise.ExercisesView.setToken"));
+    exerciseGroupsTree.getEmptyText().appendLine(
+            getText("ui.exercise.ExercisesView.setTokenDirections"));
+    emptyText.setText(getText("ui.module.ModuleListView.turnIntoAPlusProject"));
+    emptyText.setHorizontalAlignment(SwingConstants.CENTER);
+    emptyText.setVerticalAlignment(SwingConstants.CENTER);
+    mainViewModel.exercisesViewModel
+            .addValueObserver(this, ExercisesView::viewModelChanged);
```

Comment:
Sorry for once more changing my mind on this, but perhaps, now when I think of it, this would suit best to the caller method. 😞 Totally my bad. But it's that not-yet-constructed  `this` that concerns me even though it's okay here (but thinking for the future).

I only have cheek to ask this (after first saying it's fine to have it here) because I would otherwise ask you to change `viewModelChanged` private, so in any case, there would be an urge for some change. ;)